File: python/05.py
import time

# Function to perform matrix-vector multiplication
def matrix_vector_multiply(matrix, vector):
    result = [0] * len(matrix)
    for i in range(len(matrix)):
        for j in range(len(vector)):
            result[i] += matrix[i][j] * vector[j]
    return result

# Sample matrix and vector data
matrix = [[1, 2, 3],
          [4, 5, 6],
          [7, 8, 9]]
vector = [1, 2, 3]

# Start the timer
start = time.time()

# Perform matrix-vector multiplication
result = matrix_vector_multiply(matrix, vector)

# Stop the timer
end = time.time()
duration = end - start

# Display the result and the time taken
print("Result vector:", result)
print("Time taken:", duration, "seconds")

File: cpp/05.cpp
#include <iostream>
#include <vector>
#include <chrono>

// Function to perform matrix-vector multiplication
std::vector<double> matrixVectorMultiply(const std::vector<std::vector<double>>& matrix, const std::vector<double>& vector) {
    int rows = matrix.size();
    int cols = matrix[0].size();
    std::vector<double> result(rows, 0);

    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < cols; ++j) {
            result[i] += matrix[i][j] * vector[j];
        }
    }

    return result;
}

int main() {
    // Sample matrix and vector data
    std::vector<std::vector<double>> matrix = {{1, 2, 3},
                                               {4, 5, 6},
                                               {7, 8, 9}};
    std::vector<double> vector = {1, 2, 3};

    // Start the timer
    auto start = std::chrono::high_resolution_clock::now();

    // Perform matrix-vector multiplication
    std::vector<double> result = matrixVectorMultiply(matrix, vector);

    // Stop the timer
    auto end = std::chrono::high_resolution_clock::now();
    std::chrono::duration<double> duration = end - start;

    // Display the result and the time taken
    std::cout << "Result vector:";
    for (double value : result) {
        std::cout << " " << value;
    }
    std::cout << std::endl;
    std::cout << "Time taken: " << duration.count() << " seconds" << std::endl;

    return 0;
}
